onnx/helper.py

```diff
@@ -547,13 +547,14 @@ def make_optional_value_info(
         value_info_proto.doc_string = doc_string
 
     optional_type_proto = value_info_proto.type.optional_type
-    optional_type_proto.elem_type.tensor_type.elem_type = elem_type
-
-    tensor_value_info = make_tensor_value_info(name, elem_type, shape, doc_string, elem_shape_denotation)
-
-    if shape is not None:
-        optional_type_proto.elem_type.tensor_type.shape.CopyFrom(tensor_value_info.type.tensor_type.shape)
-
+    if elem_type == OptionalProto.TENSOR:
```

Comment:
I am slightly bothered by this utility: it is not fully general-purpose, and seems specialized to handle just the common cases (of the optional type). E.g., I think it won't work for creating more complex things like "optional sequence of sequence of tensor" etc., or am I mistaken? I feel it may be better to omit this utility for now (or replace with a general-purpose utility which would be extra work). Otherwise, it would be problematic to update it later on if users start taking a dependence on its current form.

Reply:
This is very close (and even more generic) than make_sequence_value_info.
This function would be needed for shape inference of an op with optional input, and I have the spec for the new op to send right after this PR.
I currently don't see any of the similar utility functions handling more complex types, and I don't also see a use-case for it.
Anyways, if you think so, I can remove it. But the following PR I have requires this.

Reply:
I agree that make_sequence_value_info has a similar problem. I am afraid that we are duplicating the problem and spreading it further. My concern is if we end up with lot of code assuming that a "sequence" is only a "sequence of tensor" or that an "optional" is only an "optional tensor" or "optional sequence of tensor" (and then we end up trying to using something more general like a "sequence of sequence of tensor" we may hit issues). When I say remove it, I mean that we can write the equivalent code in the caller. At a call-site, the caller knows that they are only creating an "optional tensor" and so write code for that, and that looks preferable. Anyway, happy to hear other opinions. I don't mind the current form if we can clean it up later, but that is tricky once users start using it widely.

Reply:
Alternatively, we could define two specialized functions with different names like "make_optional_tensor_value_info" and "make_optional_sequence_tensor_value_info" ... at least, it makes it clear it is for a special-case.